File: esp32/test/test_nearest.cpp
// Host-side unit tests for the Arduino-free ranking logic (env:native).
//   pio test -e native
#include <unity.h>

#include <cmath>
#include <vector>

#include "nearest.h"

void setUp() {}
void tearDown() {}

static Aircraft at(double lat, double lon, const char* hex) {
    Aircraft a;
    a.lat = lat;
    a.lon = lon;
    a.hex = hex;
    return a;
}

void test_radius_km_to_nm_rounds_up() {
    TEST_ASSERT_EQUAL_INT(27, radiusKmToNm(50));      // 26.99 -> 27
    TEST_ASSERT_EQUAL_INT(1, radiusKmToNm(1.852));
    TEST_ASSERT_EQUAL_INT(0, radiusKmToNm(0));
}

void test_haversine_zero_distance() {
    TEST_ASSERT_TRUE(haversineKm(50.0, 8.0, 50.0, 8.0) < 1e-9);
}

void test_haversine_known_distance() {
    // ~111 km per degree of latitude.
    double d = haversineKm(0.0, 0.0, 1.0, 0.0);
    TEST_ASSERT_TRUE(std::fabs(d - 111.19) < 0.5);
}

void test_haversine_symmetric() {
    double a = haversineKm(48.0, 2.0, 51.0, -0.1);
    double b = haversineKm(51.0, -0.1, 48.0, 2.0);
    TEST_ASSERT_TRUE(std::fabs(a - b) < 1e-9);
}

void test_rank_sorts_and_caps() {
    std::vector<Aircraft> fleet = {
        at(52.0, 8.0, "far"),
        at(50.1, 8.0, "near"),
        at(51.0, 8.0, "mid"),
    };
    rankNearest(fleet, 50.0, 8.0, 2);
    TEST_ASSERT_EQUAL_INT(2, (int)fleet.size());
    TEST_ASSERT_EQUAL_STRING("near", fleet[0].hex.c_str());
    TEST_ASSERT_EQUAL_STRING("mid", fleet[1].hex.c_str());
    TEST_ASSERT_TRUE(fleet[0].distKm < fleet[1].distKm);
    TEST_ASSERT_TRUE(fleet[0].distKm > 0.0);
}

void test_rank_negative_topn_keeps_all() {
    std::vector<Aircraft> fleet = {at(50.1, 8.0, "a"), at(52.0, 8.0, "b")};
    rankNearest(fleet, 50.0, 8.0, -1);
    TEST_ASSERT_EQUAL_INT(2, (int)fleet.size());
}

int main(int, char**) {
    UNITY_BEGIN();
    RUN_TEST(test_radius_km_to_nm_rounds_up);
    RUN_TEST(test_haversine_zero_distance);
    RUN_TEST(test_haversine_known_distance);
    RUN_TEST(test_haversine_symmetric);
    RUN_TEST(test_rank_sorts_and_caps);
    RUN_TEST(test_rank_negative_topn_keeps_all);
    return UNITY_END();
}

File: pi/tests/test_nearest.py
"""Tests for haversine distance and top-N ranking (pure math)."""

from __future__ import annotations

import math

from flight_board.nearest import bearing_degrees, haversine_km, nearest
from flight_board.source import Aircraft


def test_haversine_zero_distance():
    assert haversine_km(50.0, 8.0, 50.0, 8.0) == 0.0


def test_haversine_known_distance():
    # ~111 km per degree of latitude near the equator.
    d = haversine_km(0.0, 0.0, 1.0, 0.0)
    assert math.isclose(d, 111.19, abs_tol=0.5)


def test_haversine_symmetric():
    a = haversine_km(48.0, 2.0, 51.0, -0.1)
    b = haversine_km(51.0, -0.1, 48.0, 2.0)
    assert math.isclose(a, b, rel_tol=1e-9)


def test_bearing_degrees_cardinal_directions():
    assert math.isclose(bearing_degrees(0.0, 0.0, 1.0, 0.0), 0.0, abs_tol=0.1)
    assert math.isclose(bearing_degrees(0.0, 0.0, 0.0, 1.0), 90.0, abs_tol=0.1)
    assert math.isclose(bearing_degrees(0.0, 0.0, -1.0, 0.0), 180.0, abs_tol=0.1)
    assert math.isclose(bearing_degrees(0.0, 0.0, 0.0, -1.0), 270.0, abs_tol=0.1)


def _ac(lat: float, lon: float, hex_: str) -> Aircraft:
    return Aircraft(hex=hex_, lat=lat, lon=lon)


def test_nearest_sorts_and_caps():
    obs_lat, obs_lon = 50.0, 8.0
    fleet = [
        _ac(52.0, 8.0, "far"),
        _ac(50.1, 8.0, "near"),
        _ac(51.0, 8.0, "mid"),
    ]
    ranked = nearest(fleet, obs_lat, obs_lon, top_n=2)
    assert [a.hex for a in ranked] == ["near", "mid"]
    # Distance is filled in, single source of truth for the sort.
    assert ranked[0].dist_km < ranked[1].dist_km
    assert ranked[0].dist_km > 0
    assert ranked[0].bearing_deg is not None
    assert math.isclose(ranked[0].bearing_deg, 0.0, abs_tol=0.1)


def test_nearest_excludes_aircraft_on_ground():
    obs_lat, obs_lon = 50.0, 8.0
    fleet = [
        Aircraft(hex="ground", lat=50.0, lon=8.0, on_ground=True),
        _ac(50.1, 8.0, "airborne"),
    ]

    ranked = nearest(fleet, obs_lat, obs_lon, top_n=5)

    assert [a.hex for a in ranked] == ["airborne"]


def test_nearest_negative_top_n_returns_all():
    fleet = [_ac(50.1, 8.0, "a"), _ac(52.0, 8.0, "b")]
    ranked = nearest(fleet, 50.0, 8.0, top_n=-1)
    assert len(ranked) == 2


def test_nearest_empty():
    assert nearest([], 50.0, 8.0, top_n=5) == []
